File: js/SolverAiResultsWriter.js
const fs = require('fs');
const createCsvWriter = require('csv-writer').createArrayCsvWriter;


class SolverAiResultsWriter {
    constructor(results) {
        this.results = results;
    }

    write(path) {
        const csvWriter = createCsvWriter({
            path: path,
            append: false
        });

        const numberOfResults = this.results.getNumberOfResults();
        const inputVariableNames = this.results.getInputVariableNames();
        const outputVariableNames = this.results.getOutputVariableNames();
        const X = this.results.getX();
        const Y = this.results.getY();

        // Write headers and initial data
        csvWriter.writeRecords([
            ['Number Of Results', numberOfResults],
            ['Objective Variable Names', ...this.results.getObjectiveVariableNames()],
            ['Constraint Variable Names', ...this.results.getConstraintVariableNames()],
            [''],
            ['Inputs'],
            ['#', ...inputVariableNames]
        ]).then(() => {
            // Write input data
            for (let i = 0; i < numberOfResults; i++) {
                csvWriter.writeRecords([['X' + i, ...X[i]]]);
            }
        }).then(() => {
            // Write output headers and data
            csvWriter.writeRecords([
                [''],
                ['Outputs'],
                ['#', ...outputVariableNames]
            ]).then(() => {
                for (let i = 0; i < numberOfResults; i++) {
                    csvWriter.writeRecords([['Y' + i, ...Y[i]]]);
                }
            });
        }).then(() => {
            console.log('...Done');
        });
    }
}

module.exports = SolverAiResultsWriter;

File: python/SolverAiResultsWriter.py
import csv
import SolverAiComputeResults


class SolverAiResultsWriter:
    def __init__(self, results: SolverAiComputeResults) -> None:
        self.results = results

    def write(self, path: str) -> None:
        with open(path, 'w', newline='') as file:
            writer = csv.writer(file)
            writer.writerow(["Number Of Results",
                             self.results.getNumberOfResults()])
            writer.writerow(["Objective Variable Names"] +
                            self.results.getObjectiveVariableNames())
            writer.writerow(["Constraint Variable Names"] +
                            self.results.getConstraintVariableNames())
            writer.writerow([])

            writer.writerow(['Inputs'])
            input_names = self.results.getInputVariableNames()
            writer.writerow(['#'] + input_names)
            for i, x in enumerate(self.results.getX()):
                writer.writerow([f"X{i}"] + x)

            writer.writerow([])
            writer.writerow(['Outputs'])
            output_names = self.results.getOutputVariableNames()
            writer.writerow(['#'] + output_names)
            for i, y in enumerate(self.results.getY()):
                writer.writerow([f"Y{i}"] + y)
